rules/TypeDeclaration/Rector/FunctionLike/AddClosureParamTypeFromIterableMethodCallRector.php

```diff
@@ -0,0 +1,261 @@
+            return null;
+        }
+
+        if (! $this->callUsesClosures($node->getArgs())) {
+            return null;
+        }
+
+        $nameIndex = [];
+        foreach ($parameters as $index => $parameter) {
+            $nameIndex[$parameter->getName()] = $index;
+        }
+
+        $valueType = $varType->getIterableValueType();
+        $keyType = $varType->getIterableKeyType();
+
+        $changesMade = false;
+
+        foreach ($node->getArgs() as $index => $arg) {
+            if (! $arg instanceof Arg) {
+                continue;
+            }
+            if (! $arg->value instanceof Closure) {
+                continue;
+            }
+            $parameter = (is_string($index) ? $parameters[$nameIndex[$index]] : $parameters[$index]);
+
+            if ($this->updateClosureWithTypes($className, $parameter, $arg->value, $keyType, $valueType)) {
+                $changesMade = true;
+            }
+        }
+
+        if ($changesMade) {
+            return $node;
+        }
+
+        return null;
+    }
+
+    private function updateClosureWithTypes(
+        string $className,
+        ParameterReflection $parameter,
+        Closure $closure,
+        Type $keyType,
+        Type $valueType
+    ): bool {
+        // get the ClosureType from the ParameterReflection
+        $callableType = $this->typeUnwrapper->unwrapFirstCallableTypeFromUnionType($parameter->getType());
+
+        if (! $callableType instanceof CallableType) {
+            return false;
+        }
+
+        $changesMade = false;
+
+        foreach ($callableType->getParameters() as $index => $parameterReflection) {
+            $closureParameter = $closure->getParams()[$index] ?? null;
+
+            if (! $closureParameter instanceof Param) {
+                continue;
+            }
+
+            if (
+                $this->typeUnwrapper->isIterableTypeValue($className, $parameterReflection->getType())
+            ) {
+                if ($this->refactorParameter($closureParameter, $valueType)) {
+                    $changesMade = true;
+                }
+            } elseif (
+                $this->typeUnwrapper->isIterableTypeKey($className, $parameterReflection->getType())
+            ) {
+                if ($this->refactorParameter($closureParameter, $keyType)) {
+                    $changesMade = true;
+                }
+            }
+        }
+
+        return $changesMade;
+    }
+
+    private function refactorParameter(Param $param, Type $type): bool
+    {
+        // already set → no change
+        if ($param->type instanceof Node) {
+            $currentParamType = $this->staticTypeMapper->mapPhpParserNodePHPStanType($param->type);
+            if ($this->typeComparator->areTypesEqual($currentParamType, $type)) {
+                return false;
+            }
+        }
+
+        $paramTypeNode = $this->staticTypeMapper->mapPHPStanTypeToPhpParserNode($type, TypeKind::PARAM);
```

Comment:
`paramTypeNode` can be null, so if null, it should skip

```suggestion
        $paramTypeNode = $this->staticTypeMapper->mapPHPStanTypeToPhpParserNode($type, TypeKind::PARAM);
        if (! $paramTypeNode instanceof Node) {
            return false;
        }
```